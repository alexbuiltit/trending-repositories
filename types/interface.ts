export interface CustomError extends Error {
  displayMessage?: string;
}

export interface ErrorHandling {
  success?: boolean;
  error?: CustomError;
}
