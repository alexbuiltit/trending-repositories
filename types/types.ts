export type RepositoryType = {
  name: string;
  description?: string | null;
  starCount: number;
  id: number;
  language?: string | null;
  link: string;
  favourite?: boolean;
  toggleFavourite?: (id: number) => void;
};

export type RepositoryListType = {
  items: RepositoryType[] | undefined;
  filters?: string[] | undefined;
  toggleFavourite?: (id: number) => void;
  favourites?: number[] | undefined;
};
