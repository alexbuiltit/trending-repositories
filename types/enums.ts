export enum ListingCategory {
  All = 'all',
  Favourites = 'favourites',
}
