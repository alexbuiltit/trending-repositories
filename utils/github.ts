import { Octokit } from 'octokit';

export class Github {
  public itemsToFetch = 16;

  public async searchRepositories() {
    try {
      const octokit = new Octokit();
      const sevenDaysAgo: Date = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
      const formattedDate: string = sevenDaysAgo.toISOString().split('T')[0];

      const response = await octokit.request('GET /search/repositories', {
        sort: 'stars',
        order: 'desc',
        q: `created:>${formattedDate}`,
        per_page: this.itemsToFetch,
        page: 1,
      });

      return response.data;
    } catch (error) {
      if (error instanceof Error) {
        console.error(error);
      }
    }
  }

  public async getTrendingRepositories(count?: number) {
    if (count) {
      this.itemsToFetch = count;
    }

    const trendingRepositories = await this.searchRepositories();
    const repositoriesDataMapped = trendingRepositories?.items.map((item) => {
      return {
        name: item.name,
        description: item.description,
        starCount: item.stargazers_count,
        id: item.id,
        language: item?.language,
        link: item.html_url,
      };
    });
    return repositoriesDataMapped;
  }
}
