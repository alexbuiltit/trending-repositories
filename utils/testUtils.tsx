import React, { ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@/styles/theme';

interface ProviderProps {
  children: ReactNode;
}

const AllTheProviders = ({ children }: ProviderProps) => {
  return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: RenderOptions<typeof import('@testing-library/dom/types/queries'), HTMLElement, HTMLElement>,
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
