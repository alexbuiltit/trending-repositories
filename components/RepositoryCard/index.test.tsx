import { RepositoryType } from '@/types/types';
import { render, screen } from '@/utils/testUtils';
import { RepositoryCard } from '.';

const mockFavourite = jest.fn();

const repositoryCardProps: RepositoryType = {
  name: 'github-trending',
  description: 'A repo to display the trending items in GitHub',
  starCount: 1,
  language: 'TypeScript',
  id: 42,
  favourite: false,
  link: 'https://github.com/alexbuiltit',
  toggleFavourite: () => mockFavourite,
};

describe('RepositoryCard', () => {
  it('should render a RepositoryCard as expected', () => {
    const { container } = render(<RepositoryCard {...repositoryCardProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render a RepositoryCard that has been favourited', () => {
    repositoryCardProps.favourite = true;
    const { container } = render(<RepositoryCard {...repositoryCardProps} />);
    expect(screen.getByTestId('favourite-button')).toHaveAttribute('data-favourited', 'true');
    expect(container).toMatchSnapshot();
  });
});
