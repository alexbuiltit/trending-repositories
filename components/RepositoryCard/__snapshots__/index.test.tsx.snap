// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryCard should render a RepositoryCard as expected 1`] = `
<div>
  <li
    class="sc-c6d6bd61-0 fiYkvw"
    data-repo-id="42"
    data-testid="repository-card"
  >
    <div
      class="sc-c6d6bd61-1 hNhUMl"
    >
      <h2
        class="sc-c6d6bd61-3 cpSoRL"
      >
        <a
          href="https://github.com/alexbuiltit"
          rel="noreferrer"
          target="_blank"
        >
          github-trending
        </a>
      </h2>
      <span
        class="sc-1762ab62-0 eLoAgg"
      >
        TypeScript
      </span>
      <p>
        A repo to display the trending items in GitHub
      </p>
      <div
        class="sc-c6d6bd61-2 gychjE"
      >
        <p>
          Stars: 
          1
        </p>
        <button
          class="sc-37a9387a-0 jhFHWg"
          data-favourited="false"
          data-testid="favourite-button"
        >
          <span
            class="favourite-button__text"
          >
            Favourite
          </span>
          <span
            class="favourite-button__icon"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 24 22"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5308 3.60795L12.0001 4.88174L12.4692 3.60789C13.3602 1.18839 16.0684 0.106068 18.6509 0.62831C21.1839 1.14056 23.5 3.19824 23.5 7.00311C23.5 10.7744 20.6781 15.9886 12 21.4116C3.32188 15.9886 0.5 10.7744 0.5 7.00311C0.5 3.17441 2.81892 1.12584 5.35149 0.623033C7.93585 0.109943 10.6451 1.20372 11.5308 3.60795Z"
                stroke="black"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </li>
</div>
`;

exports[`RepositoryCard should render a RepositoryCard that has been favourited 1`] = `
<div>
  <li
    class="sc-c6d6bd61-0 fiYkvw"
    data-repo-id="42"
    data-testid="repository-card"
  >
    <div
      class="sc-c6d6bd61-1 hNhUMl"
    >
      <h2
        class="sc-c6d6bd61-3 cpSoRL"
      >
        <a
          href="https://github.com/alexbuiltit"
          rel="noreferrer"
          target="_blank"
        >
          github-trending
        </a>
      </h2>
      <span
        class="sc-1762ab62-0 eLoAgg"
      >
        TypeScript
      </span>
      <p>
        A repo to display the trending items in GitHub
      </p>
      <div
        class="sc-c6d6bd61-2 gychjE"
      >
        <p>
          Stars: 
          1
        </p>
        <button
          class="sc-37a9387a-0 jhFHWg"
          data-favourited="true"
          data-testid="favourite-button"
        >
          <span
            class="favourite-button__text"
          >
            Favourited
          </span>
          <span
            class="favourite-button__icon"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 24 22"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5308 3.60795L12.0001 4.88174L12.4692 3.60789C13.3602 1.18839 16.0684 0.106068 18.6509 0.62831C21.1839 1.14056 23.5 3.19824 23.5 7.00311C23.5 10.7744 20.6781 15.9886 12 21.4116C3.32188 15.9886 0.5 10.7744 0.5 7.00311C0.5 3.17441 2.81892 1.12584 5.35149 0.623033C7.93585 0.109943 10.6451 1.20372 11.5308 3.60795Z"
                stroke="black"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </li>
</div>
`;
